### PR TITLE
Revise section 1 of xdl

### PR DESCRIPTION
**Pull request by cheungdaven**
*Description of changes:*

By submitting this pull request, I confirm that you can use, modify,
copy, and redistribute this contribution, under the terms of your
choice.

### CODE REVIEW

**Comment by astonzhang** [comment]
Thanks! Please take a look at the CI log.

File: chapter_explainability/global_model_analysis.md [hunk 1]
```diff
@@ -0,0 +1,268 @@
...
+
+Similarly, we can remove a component (e.g., a layer or a regularizer) from the model and repeat the same procedure in order to measure the importance of that component. We leave it as an exercise.
+
+Ablation studies are straightforward and useful but they are not without shortcomings. As you may notice, the relative feature importance can be different with different ablation modes (i.e., REMOVAL and SHUFFLE). This phenomenon is quite common in explanation methods. Also, some ablation studies can be computationally expensive. Although permutation approaches can alleviate this problem, additional randomness will be introduced, that is, different runs might give different relevant feature importance.
+
+## Summary
+* Error analysis helps us get insight into the erroneous cases of the model.
+* Feature importance can be obtained via ablation studies regardless of the model type.
+
+## Exercise
```
Comment (astonzhang) [inline chapter_explainability/global_model_analysis.md:266]: Exercise -> Exercises

File: chapter_explainability/global_model_analysis.md [hunk 2]
```diff
@@ -0,0 +1,268 @@
...
+* Analyze the error distribution across different groups of the loan amount. Hint: You may need to rewrite the MlpLoan model.
```
Comment (astonzhang) [inline chapter_explainability/global_model_analysis.md:267]: Exercises have number bullets so they are easier to refer to.

File: chapter_explainability/global_model_analysis.md [hunk 3]
```diff
@@ -0,0 +1,265 @@
...
+| *last_pymnt_amnt*        | The last total payment amount received.                                           |
+| *chargeoff_within_12_mths* | The number of charge-offs within $12$ months.                               |
+
+
+We split the dataset and use $70\%$ of the examples for training and the remaining $30\%$ for testing.
+
+```{.python .input  n=9}
+#@save
+class LoanData(d2l.DataModule):
+    def __init__(self, url=d2l.DATA_URL + 'loan.csv', batch_size=64,
```
Comment (astonzhang) [inline chapter_explainability/global_model_analysis.md:37]: Since URL and LoanData are 1:1 mapping, don't treat url as an argument and hardcode it in this method.

Comment (astonzhang) [reply chapter_explainability/global_model_analysis.md:37]: Likewise, if feat_col and target_col are also determinant, hardcode them so the LoanData interface (e.g., args in the constructor) is cleaner.

File: chapter_explainability/global_model_analysis.md [hunk 4]
```diff
@@ -0,0 +1,265 @@
...
+    
+data = LoanData(url=d2l.DATA_URL + 'loan.csv', batch_size=64, test_ratio=0.3)
+```
+
+## MLPs for Loan Defaulter Prediction
+We construct a multilayer perceptron with a few fully-connected layers for loan defaulter prediction. This model is not inherently interpretable despite its simple structure.
+
+```{.python .input  n=3}
+#@save
+class MlpLoan(d2l.Classifier):
```
Comment (astonzhang) [inline chapter_explainability/global_model_analysis.md:72]: LoanMLP

File: chapter_explainability/global_model_analysis.md [hunk 5]
```diff
@@ -0,0 +1,265 @@
...
+                                 nn.ReLU(), nn.LazyLinear(32), 
+                                 nn.ReLU(), nn.LazyLinear(32), 
+                                 nn.ReLU(), nn.LazyLinear(num_outputs))
+
+    def forward(self, x):
+        return self.net(x)
+    
+    def loss(self, y_hat, y):
+        fn = nn.CrossEntropyLoss()
+        return fn(y_hat, y)
```
Comment (astonzhang) [inline chapter_explainability/global_model_analysis.md:86]: Can forward and loss in d2l.Classifier be reused here?

File: chapter_explainability/global_model_analysis.md [hunk 6]
```diff
@@ -0,0 +1,265 @@
...
+def error_analysis(self, feat_of_interest=None):
+    acc, pred, Y, X = self.black_box_model.predict(self.data.val_dataloader())
+    error = pd.DataFrame()
+    pred = torch.argmax(pred, -1)
+    error['Actual_Label'] = ["Fully-Paid" if i==1 else "Charge-Off" 
+                                    for i in Y.cpu().detach().numpy()]
+    error['Predicted_Label'] = ["Fully-Paid" if i==1 else "Charge-Off" 
+                                  for i in pred.cpu().detach().numpy()]
+    error['Status'] = ["Misclassified" if i==False else "Correctly_Classified" 
+                                for i in (pred == Y).cpu().detach().numpy()]
```
Comment (astonzhang) [inline chapter_explainability/global_model_analysis.md:143]: Is it necessary to invoke cpu()? It looks to me the data & model are already using CPU?

File: chapter_explainability/global_model_analysis.md [hunk 7]
```diff
@@ -0,0 +1,265 @@
...
+   
+    return error
+```
+
+The following function is used to draw the error maps given the features of interest.
+
+```{.python .input  n=7}
+@d2l.add_to_class(GlobalExplainer)
+def draw_heat_map(self, x_axis, y_axis):
+    d2l.plt.figure(figsize = (3,1))
```
Comment (astonzhang) [inline chapter_explainability/global_model_analysis.md:156]: `figsize = (3,1)` -> `figsize=(3, 1)`

File: chapter_explainability/global_model_analysis.md [hunk 8]
```diff
@@ -0,0 +1,265 @@
...
+    sn.heatmap(pd.crosstab(y_axis, x_axis), annot=True, cmap="GnBu", fmt='g')
```
Comment (astonzhang) [inline chapter_explainability/global_model_analysis.md:157]: Be consistent: always use single quote whenever possible

File: chapter_explainability/global_model_analysis.md [hunk 9]
```diff
@@ -0,0 +1,265 @@
...
+```
+
+### Confusion Matrix
+We can answer the first question through a confusion matrix. Confusion matrices show the inaccuracies patterns and how the model is confused when making predictions. It can provide insights into the errors being made and help identify the most promising direction to improve model performance.
+
+```{.python .input  n=8}
+feat_of_interest = {"loan_amnt": data.feat_col.index("loan_amnt"),
+               "annual_inc": data.feat_col.index("annual_inc"),
+               "term": data.feat_col.index("term"),
+               "OWN": data.feat_col.index("OWN") }
```
Comment (astonzhang) [inline chapter_explainability/global_model_analysis.md:169]: single quotes

File: chapter_explainability/global_model_analysis.md [hunk 10]
```diff
@@ -0,0 +1,265 @@
...
+            new_data = LoanData(feat_col=used_feat)
+            new_model = MlpLoan(num_outputs=2, lr=0.001, wd=1e-5)
+            trainer = d2l.Trainer(max_epochs=20)
+            trainer.fit(new_model, new_data)
+            s_i, pred, Y, X = new_model.predict(
+                new_data.val_dataloader())
+        feature_importance[i] = (s-s_i) 
+        d2l.plt.close()
+        
+    d2l.plt.figure(figsize = (3,2))
```
Comment (astonzhang) [inline chapter_explainability/global_model_analysis.md:237]: mind the space (see above comments)

File: chapter_explainability/global_model_analysis.md [hunk 11]
```diff
@@ -0,0 +1,265 @@
...
+    d2l.plt.barh(range(len(feat_of_interest)), feature_importance.values())
+    d2l.plt.yticks(range(len(feat_of_interest)), list(feature_importance.keys()))
+    d2l.plt.xlabel('Global Feature Importance score using '+mode)
```
Comment (astonzhang) [inline chapter_explainability/global_model_analysis.md:240]: add space around +

File: chapter_explainability/global_model_analysis.md [hunk 12]
```diff
@@ -0,0 +1,265 @@
...
+    d2l.plt.show()
+```
+
+The following code outputs the importance of five features: loan amount, loan term, interest rate, annual income, and the number of charge-offs within 12 months.
+
+```{.python .input  n=13}
+feat_2_analyze = ['loan_amnt', 'term', 'int_rate', 'annual_inc', 
+                  'chargeoff_within_12_mths']
+explainer.feature_importance(feat_2_analyze, mode="REMOVAL")
+explainer.feature_importance(feat_2_analyze, mode="SHUFFLE")
```
Comment (astonzhang) [inline chapter_explainability/global_model_analysis.md:250]: single quotes

File: chapter_explainability/global_model_analysis.md [hunk 13]
```diff
@@ -0,0 +1,265 @@
...
+def feature_importance(self, feat_of_interest, mode="SHUFFLE"):
+    feature_importance = {}
+    s, _, _, _ = self.black_box_model.predict(self.data.val_dataloader())
+    for i in feat_of_interest:
+        if mode == "SHUFFLE":
+            s_i, _, _, _ = self.black_box_model.predict(
+                self.data.val_dataloader(), shuffle_x=self.data.feat_col.index(i))
+        elif mode == "REMOVAL":
+            used_feat = [feat for key, feat in enumerate(self.data.feat_col) 
+                         if feat!=i]
```
Comment (astonzhang) [inline chapter_explainability/global_model_analysis.md:227]: Add space around !=

File: chapter_explainability/global_model_analysis.md [hunk 14]
```diff
@@ -0,0 +1,265 @@
...
+
+* Train the model on the entire training set and obtain an accuracy score $s$ (or other scoring metric) on the validation set.
+* For each feature $i$, $i \in m$, remove it from the training set, then retrain the model and get another score $s_i$ on the validation set. Alternatively, we can permute the feature and then re-evaluate the model :cite:`Breiman.2001`. Feature permutation is more efficient as it does not require model retraining.
+* Calculate the global feature importance score with $s-s_i$.
+
+Now, let's implement this feature ablation method. The following function provides two ablation modes: *REMOVAL* and *SHUFFLE*. To permute the feature $i$, we randomly shuffle the values of feature $i$ in the current batch.
+
+```{.python .input  n=12}
+@d2l.add_to_class(GlobalExplainer)
+def feature_importance(self, feat_of_interest, mode="SHUFFLE"):
```
Comment (astonzhang) [inline chapter_explainability/global_model_analysis.md:218]: If `feat_of_interest` is a list, plural naming enhances readability: `features_of_interest`

File: chapter_explainability/global_model_analysis.md [hunk 15]
```diff
@@ -0,0 +1,265 @@
...
+            error[key] = X[:, value].cpu().detach().numpy()  
+   
+    return error
+```
+
+The following function is used to draw the error maps given the features of interest.
+
+```{.python .input  n=7}
+@d2l.add_to_class(GlobalExplainer)
+def draw_heat_map(self, x_axis, y_axis):
```
Comment (astonzhang) [inline chapter_explainability/global_model_analysis.md:155]: Should this function be added to GlobalExplainer as its method? In other words, is `self` used in this function? Will this function be used by other explainers?

**Review by astonzhang — COMMENTED** [review]
Some of my comments may be applicable to other places (e.g., single quotes). Please apply them wherever possible.

File: chapter_explainability/intro.md [hunk 16]
```diff
@@ -59,6 +56,6 @@ We will lay out this chapter by first introducing global explanation methods (e.
 * Explanations can be on the entire model or at the instance level.
 * Perils and pitfalls exist in explanation methods.
 
-## Excersie
-* Can you name some scenarios where explanations are critical?
-* When should we use post-hoc explanation methods rather than self-explanatory methods?
+## Excersies
```
Comment (astonzhang) [inline chapter_explainability/intro.md:59]: Exercises

**Comment by astonzhang** [comment]
Merging this to allow for Dominik's edits. Will iterate it over.
